src/App.js

```diff
@@ -29,37 +29,101 @@ const App = () => {
+  const [winner, setWinner] = useState(null);
+  const [gameEnd, setGameEnd] = useState(false);
 
-  // Wave 2
-  // You will need to create a method to change the square
-  //   When it is clicked on.
-  //   Then pass it into the squares as a callback
+  const currentPlayer = (turn) => {
+    if (turn <= 9) {
+      if (turn % 2 !== 0) {
+        return PLAYER_1;
+      } else {
+        return PLAYER_2;
+      }
+    }
+  };
+
+  const onClickCallback = (id) => {
+    const updateSquare = squares.map((square) => {
+      while (gameEnd === false) {
+        for (let subSquare of square) {
+          if (subSquare.id === id && subSquare.value === '') {
+            subSquare.value = currentPlayer(turn);
+            if (turn < 9) {
+              setTurn(turn + 1);
+            }
+          }
+          let checkedWinner = checkForWinner();
+          console.log(checkedWinner);
+          if (checkedWinner !== null) {
+            setWinner(checkedWinner);
+            setGameEnd(true);
+            break;
+          }
+        }
+        return square;
+      }
+      setSquares(updateSquare);
+    });
+  };
 
   const checkForWinner = () => {
-    // Complete in Wave 3
-    // You will need to:
-    // 1. Go accross each row to see if
-    //    3 squares in the same row match
-    //    i.e. same value
-    // 2. Go down each column to see if
-    //    3 squares in each column match
-    // 3. Go across each diagonal to see if
-    //    all three squares have the same value.
+    let i = 0;
+
+    // Check all the rows and columns for a winner
+    while (i < 3) {
+      if (
+        squares[i][0].value === squares[i][1].value &&
+        squares[i][2].value === squares[i][1].value &&
+        squares[i][0].value !== ''
+      ) {
+        return squares[i][0].value;
+      } else if (
+        squares[0][i].value === squares[1][i].value &&
+        squares[2][i].value === squares[1][i].value &&
+        squares[0][i].value !== ''
+      ) {
+        return squares[0][i].value;
+      }
+      i += 1;
+    }
+    // Check Top-Left to bottom-right diagonal
+    if (
+      squares[0][0].value === squares[1][1].value &&
+      squares[2][2].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][0].value;
+    }
+
+    // Check Top-right to bottom-left diagonal
+    if (
+      squares[0][2].value === squares[1][1].value &&
+      squares[2][0].value === squares[1][1].value &&
+      squares[1][1].value !== ''
+    ) {
+      return squares[0][2].value;
+    }
+
+    return null;
   };
 
   const resetGame = () => {
-    // Complete in Wave 4
+    setSquares(generateSquares());
+    setTurn(1);
+    setWinner(null);
+    setGameEnd(false);
   };
 
   return (
     <div className="App">
       <header className="App-header">
         <h1>React Tic Tac Toe</h1>
-        <h2>The winner is ... -- Fill in for wave 3 </h2>
-        <button>Reset Game</button>
+        <h2>Winner is {winner}</h2>
```

Comment:
When there isn't a winner you could display who the current player is. Like "Its Player 1's turn" or "It's X's turn"